### DISCUSSION TITLE
parallel reading polyhedral meshes

### DISCUSSION THREAD

**Discussion by jaapGit** [CGNS/CGNS#732] [CGNStalk]
Hi all,

I am currently working on reading and writing cgns data in parallel using the cgns lib. Until now we where reading the data on the master proc and distributed it to the other procs but this has its limits with respect to memory usage and performance. Is it, using the latest cgns version, possible to read and write polyhedral meshes in parallel? It seems that some cgns routines are missing:

cgp_poly_elements_read_data_f

routines to read and write the parent data

Or is my understanding incorrect?

Best regards

Jaap Windt

**Comment by brtnfld** [comment]
I'm working on the missing Fortran APIs here:

https://github.com/brtnfld/CGNS/tree/CGNS-276

[ref] https://github.com/CGNS/CGNS/issues/671

**Reply by brtnfld** [reply]
Actually, there is no cgp_poly_elements_read_data in general.

**Reply by jaapGit** [reply]
No, and also nothing to read/write parent data, right? (it is available in CGNS-276, though)

**Reply by jrwrigh** [reply]
There is a parallel parent data writer on `develop` right now. See this test for usage (I'm not sure if it's documented right now): https://github.com/CGNS/CGNS/blob/develop/src/ptests/test_unstruc_quad.c

**Reply by jaapGit** [reply]
I have been playing a bit with the fortran interface, maybe it is helpful. I really would like to test the parallel reading/writing of polyhedral grids this year. I might contribute writing the fortran interface and testing (Lustre). And with polyhedral I mean grids which do not have constant size elements. I guess we need a route to read the offsets first and then a routine to read the element data. Is this realistic or is there too much to develop yet and are these developments going much slower? Besides my concern about the non constant size elements I am concerned about the memory usage when reading poly grids because I thought I read somewhere that for poly grids the complete elements data stays in memory. Or is this nonsense? 
 
[cgns_f.4.1.2.txt.patch](https://github.com/CGNS/CGNS/files/12461025/cgns_f.4.1.2.txt.patch)

**Reply by jaapGit** [reply]
Could someone please explain what the problem with true polyhedra grids is. Wil parallel-io be supported in the future and what is the expected, approximate, time for this to be complete (end 2023, 2024, 2025?) Can we somehow contribute to these developments next year?

**Reply by MicK7** [reply]
What is missing from your point of view and need to be developped ? 
For reading at least, we already have parallel feature by using cg_ElementPartialSize and cg_poly_elements_partial_read in solver like SU2. What else is required ?

**Reply by jaapGit** [reply]
cg_poly_elements_partial_read is not for reading in parallel right? I used it to read the cpu specific part in serial but that is much too slow. Besides that the memory usage when using cg_poly_elements_partial_read is huge. Probably the complete data set stays in memory. But that is what I know now. There might be good reasons why the library works as it is now. Hence my general question. Kind regards, Jaap (j.windt@marin.nl)

**Reply by MicK7** [reply]
If you are using CGNS standard 4 files, the two functions should not load the complete data in memory but only the requested part.
The function works in serial but can also be used in parallel if the file is open in parallel.

**Reply by jaapGit** [reply]
Hmm, maybe I did something wrong then. As far as I know I used standard 4 files. I will look into it again and will redo my valgrind massif tests as soon as I have time. Thanks for the info.

**Reply by jaapGit** [reply]
With respect to Element connectivity the midlevel library mentions: "In most cases, the data is not duplicated in the mid-level library and written directly from the user data to disk.The exception to this is in the case of MIXED, NGON_n, and NFACE_n element sets. Even if the size of the element connectivity array can be known through ConnectOffset, the MLL will keep a copy of the data in memory for the partial writes." 

This suggests to confirm my valgrind massif test, right?

**Reply by MicK7** [reply]
Indeed after checking the code, the offset array is fully stored in memory but the elements are not. So you are right there is still room for improvement.

**Reply by MicK7** [reply]
My conclusion were for the reading part. For the writing part, you need to use cgp_poly_section_write followed by cgp_poly_elements_write_data ... they should not store the whole data ...  but as you mentionned valgrind test is needed to check that.

**Reply by jaapGit** [reply]
Thanks for the response.

For my goal it is a problem that the offset array is fully stored. For now we read the complete data and scatter it to the processes. The same is done for writing. This is fast but for the future it is not good enough because it uses to much memory. Therefore I want to try reading in parallel. If the offset is fully stored the memory problems stays, although it is a bit smaller (which is not my impression from valgrind). 

The cgp_poly_elements_write_data routines do not work with NFACE_n data: "These routines only work for constant sized elements, since it is not possible to compute file offsets for variable sized elements without knowledge of the entire element connectivity data" which is indeed confirmed when I tried them. 

By the way, you mentioned that all the cg_ routines can be used in parallel too (compiling the lib with DCGNS_ENABLE_PARALLEL:BOOL=ON and use cgp_open, right?). I did not know that, good to hear, thanks. I thought that one had to use the cgp_ routines for parallel reading of coordinates, elements data, etc. I am not sure but I think it was like that in the past. So the cgp_ routines are not needed expect when you want to use Lustre for example? 

Something else: If the number of processes is larger than the number of elements in the data set (this happens with unstructured grids easily for one of the boundaries) the parallel reading fails because the requested size (zero) is invalid. It was also not possible, obviously, to skip the reading on some processes because then the program hangs.

best regards,

Jaap
